baseband/gsb/tests/test_gsb.py

```diff
@@ -270,241 +331,286 @@ def test_rawdump_frame(self, tmpdir):
+            assert fh_r.header0.time == fh_r.start_time
+            assert fh_r.header0 == frame1.header
+            assert np.all(fh_r.read(fh_r.samples_per_frame) == frame1.data)
+            for pol in fraw:
+                for thread in pol:
+                    thread.close()
+
+        # Try writing to file by passing header keywords into open.
+        with gsb.open(SAMPLE_PHASED_HEADER, mode='rs', raw=SAMPLE_PHASED,
+                      samples_per_frame=2**3) as fh_r, \
+                io.open(str(tmpdir.join('test_time.timestamp')), 'w+b') as sh,\
                 open(str(tmpdir.join('test0.dat')), 'w+b') as sp0, \
                 open(str(tmpdir.join('test1.dat')), 'w+b') as sp1, \
                 open(str(tmpdir.join('test2.dat')), 'w+b') as sp2, \
-                open(str(tmpdir.join('test3.dat')), 'w+b') as sp3, \
-                gsb.open(sh, 'ws', raw=((sp0, sp1), (sp2, sp3)),
-                         bps=bps, sample_rate=128*u.Hz,
-                         samples_per_frame=twopol.shape[0] // 2,
-                         nchan=twopol.shape[2], nthread=twopol.shape[1],
-                         complex_data=True, header=header) as fh_w:
-            # Write data twice.
-            assert fh_w.sample_shape == (2, twopol.shape[2])
-            fh_w.write(twopol)
-            fh_w.write(twopol[::-1])
-            assert fh_w.tell() == twopol.shape[0] * 2
-            assert_quantity_allclose(fh_w.tell(unit=u.s), 0.5 * u.s)
-            assert fh_w._header['seq_nr'] == fh_w.header0['seq_nr'] + 3
+                open(str(tmpdir.join('test3.dat')), 'w+b') as sp3:
+            fraw = ((sp0, sp1), (sp2, sp3))
+            fh_w = gsb.open(sh, 'ws', raw=fraw, samples_per_frame=2**3,
+                            frames_per_second=fh_r.frames_per_second,
+                            **fh_r.header0)
+            fh_w.write(fh_r.read())
             fh_w.flush()
-            assert sp0.tell() == 1024 * bps // 8
-            for fh in sh, sp0, sp1, sp2, sp3:
-                fh.seek(0)
-            with gsb.open(sh, mode='rs', raw=((sp0, sp1), (sp2, sp3)),
-                          bps=bps, samples_per_frame=twopol.shape[0] // 2,
-                          nchan=twopol.shape[2]) as fh_r:
-                assert fh_r.header0 == header
-                assert np.isclose(fh_r.frames_per_second, 8.)
-                data = fh_r.read(twopol.shape[0] * 2)
-                assert fh_r.tell() == twopol.shape[0] * 2
-                assert (fh_r._frame.header['seq_nr'] ==
-                        fh_r.header0['seq_nr'] + 3)
-                assert_quantity_allclose(fh_r.tell(unit=u.s), 0.5 * u.s)
-                assert sp0.tell() == 1024 * bps // 8
-                assert fh_r._frame.header == fh_r._last_header
-            assert np.all(data[:twopol.shape[0]] == twopol)
-            assert np.all(data[twopol.shape[0]:] == twopol[::-1])
+            sh.seek(0)
+            for pol in fraw:
```

Comment:
now seen this so many times that I wonder if the test class cannot use a little helper function!